text/0000-bench-utils.md

```diff
@@ -0,0 +1,157 @@
+- Feature Name: black_box-and-clobber
+- Start Date: 2018-03-12
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC adds two functions to `core::mem`: `black_box` and `clobber`, which are
+mainly useful for writing benchmarks.
+
+# Motivation
+[motivation]: #motivation
+
+The `black_box` and `clobber` functions are useful for writing synthetic
+benchmarks where, due to the constrained nature of the benchmark, the compiler
+is able to perform optimizations that wouldn't otherwise trigger in practice.
+
+The implementation of these functions is backend-specific and requires inline
+assembly. Such that if the standard library does not provide them, the users are
+required to use brittle workarounds on nightly.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+
+## `mem::black_box`
+
+The function:
+
+```rust
+pub fn black_box<T>(x: T) -> T;
+```
+
+prevents the value `x` from being optimized away and flushes pending reads/writes
+to memory. It does not prevent optimizations on the expression generating the
+value `x` nor on the return value of the function. For
+example ([`rust.godbolt.org`](https://godbolt.org/g/YP2GCJ)):
+
+```rust
+fn foo(x: i32) -> i32{ 
+  mem::black_box(2 + x);
+  3
+}
+let a = foo(2);
+```
+
+Here, the compiler can simplify the expression `2 + x` into `2 + 2` and then
+`4`, but it is not allowed to discard `4`. Instead, it must store `4` into a
+register even though it is not used by anything afterwards.
+
+## `mem::clobber`
+
+The function
+
+```rust
+pub fn clobber() -> ();
+```
+ 	 
+flushes all pending writes to memory. Memory managed by block scope objects must
```

Comment:
This wording (flushing pending writes) makes me uncomfortable because it's remniscient of  memory consistency models,  including hardware ones, when this is just a single-threaded and compiler-level restriction. Actually, come to think of it, I'd like to know the difference between this and [`compiler_fence(SeqCst)`](https://doc.rust-lang.org/std/sync/atomic/fn.compiler_fence.html). I can't think of any off-hand.

Reply:
> when this is just a single-threaded and compiler-level restriction

This is correct. 

> compiler_fence(SeqCst). I can't think of any off-hand.

I can't either, but for some reason they do generate different code: https://godbolt.org/g/G2UoZC

I'll give this some more thought.

Reply:
The difference between `asm!` with a `memory` clobber and `compiler_fence` exists in that, that `memory` clobber requires compiler to actually reload the memory if they want to use it again (as memory is… clobbered – considered changed), whereas `compiler_fence` only enforces that memory accesses are not reordered and the compiler still may use the usual rules to figure that it needn’t to reload stuff.

Reply:
@nagisa the only thing that clobber should do is flush pending writes to memory. It doesn't need to require that the compiler reloads memory on reuse. Maybe the `volatile` `asm!` with  `memory` clobber is not the best way to implement that.

Reply:
@nagisa Thank you. I was mislead by the fact that fences prohibit some load-store optimization to thinking they'd also impact things like store-load forwarding on the same address with no intervening writes.

@gnzlbg `asm!` with `memory` is simply assumed to read from and write to all memory and all its effects follow from that. However, to be precise, this does not mean any reloads are *introduced* after a clobbering inline asm, it just means that all the loads that is already there (of which are a lot given that every local and many temporaries are stack slots) can't be replaced with values loaded from the same address before the clobber.

If you want something less strong, you need to be precise. "Flushing loads" is not really something that makes intuitive sense at the compiler level (and it surely has no effect on runtime, for example caches or store buffers?).

Reply:
@rkruppe `mem::clobber()` should be assumed to read from all memory with effects, so that any pending memory stores must have completed before the `clobber`. All the loads that are already there in registers, temporary stack slots, etc, should not be invalidated by the `clobber`.

Reply:
Okay that is a coherent concept. I'm not sure off-hand how to best implement that in LLVM. (`compiler_fence` is probably not enough, since it permits dead store elimination if the memory location is known to not escape.)

Reply:
However, come to think of it, what's the difference between `black_box(x)` (which is currently stated to just force a write of `x` to memory) and `let tmp = x; clobber();` (which writes `x` to the stack slot of `tmp` and then forces that store to be considered live)?

Reply:
That's a good question and this relates to what I meant with "block scope" . In

```rust
{ 
    let tmp = x; 
    clobber();
}
```

this `{}` block is the only part of the program that knows the address of `tmp` so no other code can actually read it without invoking undefined behavior. So in this case, `clobber` does not make the store of `tmp` live, because nothing is able to read from it.

However, if one shares the address of `tmp` with the rest of the program:

```rust
{ 
    let tmp = x; 
    black_box(&tmp);
    clobber();
}
```

then `clobber` will force the store of `tmp` to be considered live.